File: order/quicksort/java/QuickSort.java
package io.github.oxnz.algorithms.quicksort;

public class QuickSort {
    private int[] numbers;
    private int number;

    public void sort(int[] values) {
        // check for empty or null array
        if (values == null || values.length == 0) { return; }
        this.numbers = values;
        number = values.length;
    }

    private void quicksort(int low, int high) {
        int i = low, j = high;
        // Get the pivot element from the middle of the list
        int pivot = numbers[i+(j-i)/2];

        // Divid into two lists
        while (i <= j) {
            // If the current value from the left list is smaller than
            // the pivot element then get the next element from the left
            // list
            while (numbers[i] < pivot) { ++i; }
            // If the current value from the right list is larger than
            // the pivot element, then get next element from the right
            // list
            while (numbers[j] > pivot) { --j; }

            // If we have found a values in the left list which is larger
            // than the pivot element and if we have found a value in the
            // right list which is smaller than the pivot element, then
            // we exchange the values.
            // As we are done we can increase i and j
            if (i <= j) { exchange(i++, j--); }
        }
        // Recursion
        if (low < j) quicksort(low, j);
        if (i < high) quicksort(i, high);
    }

    private void exchange(int i, int j) {
        int temp = numbers[i];
        numbers[i] = numbers[j];
        numbers[j] = temp;
    }
}

File: order/quicksort/cpp/quicksort.cpp
#include <iostream>

using namespace std;

template<typename T, typename Index>
void quicksort(T array[], Index left, Index right) {
    if (left < right) {
        Index l = left;
        Index r = right;
        T m = array[l];
        while (l < r) {
            while (l < r && array[r] >= m)
                --r;
            if (l < r)
                array[l++] = array[r];
            while (l < r && array[l] < m)
                ++l;
            if (l < r)
                array[r--] = array[l];
        }
        array[l] = m;
        quicksort(array, left, l-1);
        quicksort(array, l+1, right);
    }
}

template <typename T, typename Index>
void printArray(T a[], Index maxIndex) {
    for (Index i = 0; i < maxIndex; ++i) {
        cout << a[i] << " ";
    }
    cout << endl;
}

int main() {
    int a[20] = {0, 4, 332232, 434, 5454, 423, 7, 32132131, 4343, 434};
    printArray(a, 20);
    quicksort(a, 0, 19);
    printArray(a, 20);
    
    return 0;
}
